Modules/Core/Common/include/itkMath.h

```diff
@@ -318,25 +319,21 @@ FloatAlmostEqual(T                                        x1,
     return true;
   }
 
-#if defined(__APPLE__) && (__clang_major__ == 3) && (__clang_minor__ == 0) && defined(NDEBUG) && defined(__x86_64__)
-  Detail::FloatIEEE<T> x1f(x1);
-  Detail::FloatIEEE<T> x2f(x2);
-  double               x1fAsULP = static_cast<double>(x1f.AsULP());
-  double               x2fAsULP = static_cast<double>(x2f.AsULP());
-  double               ulps = x1fAsULP - x2fAsULP;
-  if (ulps < 0)
+  // This check for different signs is necessary for several reasons, see the blog link above.
+  // Subtracting the signed-magnitude representation of floats using twos-complement
+  // math isn't particularly meaningful, and the subtraction would produce a 33-bit
+  // result and overflow an int.
+  if (std::signbit(x1) != std::signbit(x2))
```

Comment:
This looks like a major bug fix to me, thanks again Sean. Apparently this essential sign comparison was missing already from the very first version by @thewtex and @bradking, Feb 25, 2012: 0d8ca651d64e8c03efbdb807ed8d2f5893dc8ee5

Reply:
It's funny, because when I was debugging this, I thought: "I remember reading a great blog about this", then was surprised to see the blog mentioned in the ITK comments!  But for some reason the blogger's implementation was not copied fully. ITK is now doing basically exactly the same as his `AlmostEqualUlpsAndAbs` function. (Which BTW excludes the 0 == -0 part, for the reason you surmised).

It was sorta caught in late 2015 with 207a801, but the change only hid the problem really.

Floating point math is hard. :)

Reply:
The implementation deviated, e.g. adding support for both `float` and `double`. But, this comparison was lost in the process. Lucky us, we now have `std::signbit` we can use with C++11 :-).